actions/contacts/upsert.go

```diff
@@ -0,0 +1,63 @@
+package contacts
+
+import (
+	"encoding/json"
+	"errors"
+	"net/http"
+
+	"github.com/bitcoin-sv/spv-wallet/engine"
+	"github.com/bitcoin-sv/spv-wallet/mappings"
+	"github.com/bitcoin-sv/spv-wallet/models"
+	"github.com/bitcoin-sv/spv-wallet/server/auth"
+	"github.com/gin-gonic/gin"
+)
+
+// upsert will add new or modified existing contact
+// Upsert contact godoc
+// @Summary		Upsert contact
+// @Description	Add or update contact. For new contact send request to add current user as contact
+// @Tags		Contact
+// @Produce		json
+// @Param		UpsertContact body contacts.UpsertContact true "Full name and metadata needed to add/modify contact"
+// @Success		201
+// @Router		/v1/contact/{paymail} [PUT]
+// @Security	x-auth-xpub
+func (a *Action) upsert(c *gin.Context) {
+	requesterPubKey := c.GetString(auth.ParamXPubKey)
+	cPaymail := c.GetString("paymail")
+
+	var req UpsertContact
+	if err := json.NewDecoder(c.Request.Body).Decode(&req); err != nil {
```

Comment:
Consider using one of Gin BindJson function, which also provides simple validation.

https://gin-gonic.com/docs/examples/binding-and-validation/

But I don't have any problem with the standard json-decoder approach.

Reply:
Good suggestion, especially if it supports payload validation. However, I would prefer to change the entire system with a rethink of returned statuses.

Reply:
Good point 💯 When the refactorization time comes we could discuss it and make it consistent